runtime/auth/aws-signing-common/common/src/aws/smithy/kotlin/runtime/auth/awssigning/middleware/AwsSigningMiddleware.kt

```diff
@@ -147,17 +149,28 @@ public class AwsSigningMiddleware(private val config: Config) : ModifyRequestMid
                 contextHashSpecification != null -> contextHashSpecification
                 config.isUnsignedPayload -> HashSpecification.UnsignedPayload
                 body is HttpBody.Empty -> HashSpecification.EmptyBody
-                body is HttpBody.Streaming && !body.isReplayable -> {
-                    coroutineContext.warn<AwsSigningMiddleware> {
-                        "unable to compute hash for unbounded stream; defaulting to unsigned payload"
+                body is HttpBody.Streaming && (!body.isReplayable || body.isEligibleForAwsChunkedStreaming()) -> {
+                    if (req.subject.headers.contains("x-amz-trailer")) {
+                        HashSpecification.StreamingAws4HmacSha256PayloadWithTrailers
+                    } else {
+                        HashSpecification.StreamingAws4HmacSha256Payload
                     }
-                    HashSpecification.UnsignedPayload
                 }
                 // use the payload to compute the hash
                 else -> HashSpecification.CalculateFromPayload
             }
         }
 
+        when (signingConfig.hashSpecification) {
+            // Add aws-chunked headers
+            is HashSpecification.StreamingAws4HmacSha256Payload, is HashSpecification.StreamingAws4HmacSha256PayloadWithTrailers -> {
+                req.subject.headers.append("Content-Encoding", "aws-chunked")
+                req.subject.headers.append("Transfer-Encoding", "chunked")
+                req.subject.headers.append("X-Amz-Decoded-Content-Length", (body.contentLength ?: 0).toString())
+            }
+            else -> {}
+        }
```

Comment:
**Question**: It seems odd to do this outside of the signer/canonicalizer. Why is this the right place to mutate the request for chunked encoding?

Reply:
I agree, it seems out of place when comparing to the canonicalizer. I just finished played around with it and don't think appending these headers should happen in the canonicalizer, because only the default signer uses it -- the CRT signer canonicalizes requests internally.

Do you think creating a new middleware here makes sense? Something that will check if the body should be uploaded with `aws-chunked`, and append the appropriate headers if so. 

Also, what do you think about adding a new middleware for lines 183-191 below, where the request body is wrapped in an `AwsChunkedByteReadChannel`? It also seems a little out of place here in the signing middleware. One challenge with a new middleware would be getting access to the signing config -- it would have to be added to the execution context here and fetched in the new middleware.

Reply:
This isn't part of canonicalization though so it doesn't belong there. It's part of `aws-chunked` content encoding. We make a decision to encode it a certain way and need to update the body to reflect that decision. 

I don't think it belongs in a separate middleware (nor the other lines below).